metal/resource_metal_port.go

```diff
@@ -0,0 +1,342 @@
+ - native VLAN detachment
+ - VLANs detachment
+ - bond break
+ - l2/l3 conversion
+ - bond creation
+ - VLANs assignment
+ - native VLAN assignment
+*/
+
+func resourceMetalPortCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+
+	port, err := getPortByResourceData(d, client)
+	if err != nil {
+		return err
+	}
+
+	// Constraint: Only bond ports have layer2 mode
+	l2, l2Ok := d.GetOk("layer2")
+	if port.Type != "NetworkBondPort" && l2Ok {
+		return fmt.Errorf("layer2 flag can be specified only for bond ports")
+	}
+
+	// remove native vlan
+	currentNative := getCurrentNative(port)
+	specifiedNative := getSpecifiedNative(d)
+	if (currentNative != specifiedNative) && currentNative != "" {
+		_, _, err = client.DevicePorts.UnassignNative(port.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	// remove vlans
+	vlansToRemove := difference(
+		attachedVlanIds(port),
+		specifiedVlanIds(d),
+	)
+	err = processVlansOnPort(port.ID, vlansToRemove, client.DevicePorts.Unassign)
+	if err != nil {
+		return err
+	}
+
+	// disbond
+	bonded, bondedOk := d.GetOk("bonded")
+	if bondedOk && !bonded.(bool) && port.Data.Bonded {
+		port, _, err = client.DevicePorts.Disbond(port, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	// convert to layer2
+	isLayer2 := contains(l2Types, port.NetworkType)
+
+	if l2Ok && l2.(bool) && !isLayer2 {
+		port, _, err = client.Ports.ConvertToLayerTwo(port.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	// convert to layer3
+	if l2Ok && !l2.(bool) && isLayer2 {
+		ips := []packngo.AddressRequest{
+			{AddressFamily: 4, Public: true},
+			{AddressFamily: 4, Public: false},
+			{AddressFamily: 6, Public: true},
+		}
+		port, _, err = client.Ports.ConvertToLayerThree(port.ID, ips)
+		if err != nil {
+			return err
+		}
+	}
+
+	// constructive phase - bond creation and vlan assignment
+	// ======================================================
+
+	port, err = getPortByResourceData(d, client)
+	if err != nil {
+		return err
+	}
+
+	// bond
+	if bondedOk && bonded.(bool) && !port.Data.Bonded {
+		port, _, err = client.DevicePorts.Bond(port, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	// assign VLANs
+	vlansToAssign := difference(
+		specifiedVlanIds(d),
+		attachedVlanIds(port),
+	)
+	err = processVlansOnPort(port.ID, vlansToAssign, client.DevicePorts.Assign)
+	if err != nil {
+		return err
+	}
+
+	// assign Native VLAN
+	currentNative = getCurrentNative(port)
+	specifiedNative = getSpecifiedNative(d)
+	if (currentNative != specifiedNative) && currentNative != "" {
+		par := packngo.PortAssignRequest{
+			PortID:           port.ID,
+			VirtualNetworkID: specifiedNative,
+		}
+		_, _, err = client.DevicePorts.AssignNative(&par)
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceMetalPortRead(d, meta)
+}
+
+func resourceMetalPortRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+	port, err := getPortByResourceData(d, client)
```

Comment:
Does it seem unusual for a resource to load from either id or a composite key (name,device_id)? It is usual for data sources to work this way, but I wonder if this is safe on a resource.

Reply:
I was not safe and I removed it. metal_port now needs only port id